### PR TITLE
Refactor running of interactive commands

### PR DESCRIPTION
**Pull request by happz**
* more unified implementation,
* raises `RunError` when command fails, dealing with the fallout is left to the caller,
* `login` may ignore it, `execute/internal` will not - interactive tests apparently could not fail.

Pull Request Checklist

* [x] implement the feature
* [x] extend the test coverage

### CODE REVIEW

**Comment by LecrisUT** [comment]
Today I've found that there is a standard for debugger-ide interface: https://microsoft.github.io/debug-adapter-protocol/. Saw it because CMake recently implemented it. Any ideas if it can fit around this, or would it need a lot more backend?

**Comment by thrix** [comment]
> Today I've found that there is a standard for debugger-ide interface: https://microsoft.github.io/debug-adapter-protocol/. Saw it because CMake recently implemented it. Any ideas if it can fit around this, or would it need a lot more backend?

@LecrisUT not sure I follow your thoughts, you would like to be able to debug (step by step) the test execution somehow via vscode interface? :) It should be possible now to drop to an interactive shell if you run tmt in the vscode shell right? (I wonder what I am not understanding)

**Comment by happz** [comment]
> Today I've found that there is a standard for debugger-ide interface: https://microsoft.github.io/debug-adapter-protocol/. Saw it because CMake recently implemented it. Any ideas if it can fit around this, or would it need a lot more backend?

No idea, TBH, but I'm tempted to say "not really". The patch tries to address how tmt runs "interactive" commands, so it'd be no longer different than how non-interactive commands are handled, especially when it comes to exceptions and return values. I suppose anything as sophisticated as debugging protocols is not related to this particular PR.

If what you're thinking about is something like running a test with a debugging session open elsewhere, to your IDE or gdb in another SSH session, e.g. for runtime inspection of the debugged test process, then we should probably look into test checks. Those should enhance or spice up the test process, by checking related events (`abrt`, AVC denials raised during the test, etc.), but I can imagine also some kind of "middleware" check that would do something - open this debugger-IDE session - and then proceed running the test as usual. Did you mean something along these lines?

**Comment by LecrisUT** [comment]
I think the protocol is more geared towards providing the IDEs the data model they need for interacting with the debugger, e.g. setting environment variables, context, etc.

Currently the scope of this PR is to get the interactive shell, maybe analogous to the ansible debugger shell, where you have to actively step through each step and interact with the shell environment?

**Comment by happz** [comment]
> Currently the scope of this PR is to get the interactive shell, maybe analogous to the ansible debugger shell, where you have to actively step through each step and interact with the shell environment?

Yes and no. It is aiming at this use case, i.e. when a user adds `/bin/bash` into their test, running the test in "interactive" mode would mean this spawned shell would be accessible to the user. But, also no in the sense that it is not a new feature added by this PR, it's been present already. This PR tries to fix some issues discovered in December, namely the failure of a test in interactive mode is ignored. Which is acceptable with the `login` subcommand, but not with regular tests. Plus there were logging issues, some steps the code takes in the case of the non-interactive mode would not happen in the interactive mode, and so on.

In other words, you got the right idea, but I'm not adding it, just trying to fix a few bugs and unify implementation to make future reasoning about the interactive mode easier :)

**Review by psss — APPROVED** [review]
Looks good and works nice, thanks for improving this!

**Comment by psss** [comment]
Seems that `/tests/execute/duration` will now need more time for the extra loop (timeouted) --> increased.

**Comment by psss** [comment]
> Seems that `/tests/execute/duration` will now need more time for the extra loop (timeouted) --> increased.

Actually, when run in the interactive mode the duration is ignored which means the long test waits 120 seconds doing nothing. Do we really want/need to exercise this scenario?

**Comment by psss** [comment]
> I'd say we can make it shorter, e.g. 60 seconds or something, although I'd advise against being stingy & cutting it too close to 5 seconds...

Sounds ok. I'd just go a bit lower, what about 20 seconds? It's still 4 times more that the `duration`. Should be safe, hm? I would like to avoid slowing down the test suite unnecessarily.

**Comment by happz** [comment]
> > I'd say we can make it shorter, e.g. 60 seconds or something, although I'd advise against being stingy & cutting it too close to 5 seconds...
> 
> Sounds ok. I'd just go a bit lower, what about 20 seconds? It's still 4 times more that the `duration`. Should be safe, hm? I would like to avoid slowing down the test suite unnecessarily.

TBH, I can't tell. Nobody can. Feel free to lower it down to 20 seconds, it's probably going to work as well as it used to. It's probably not much worse than 21 seconds, or 30 seconds. I simply have a bad experience with this kind of optimization, let's finish the discussion here.

**Comment by psss** [comment]
Ok, adjusted in 3b85b370. Should save us ~16 minutes of waiting.

**Comment by psss** [comment]
Rawhide failures are irrelevant --> merging.
